syntax = "proto3";

package SimpleBackendServiceCppGrpc;

service JobService {
  rpc SubmitJob (SubmitJobRequest) returns (SubmitJobResponse);
  rpc GetJobStatus (GetJobStatusRequest) returns (GetJobStatusResponse);
  rpc GetJobResult (GetJobResultRequest) returns (GetJobResultResponse);
}

enum JobStatus {
  JOB_STATUS_UNKNOWN = 0;
  JOB_STATUS_PENDING = 1;
  JOB_STATUS_RUNNING = 2;
  JOB_STATUS_COMPLETED = 3;
  JOB_STATUS_FAILED = 4;
  JOB_STATUS_NO_RESULT_FOUND = 5; // This should never happen under normal circumstances
}

message SubmitJobRequest {
  string type = 1;
  string payload = 2;
}

message SubmitJobResponse {
  uint64 job_id = 1;
}

message GetJobStatusRequest {
  uint64 job_id = 1;
}

message GetJobStatusResponse {
  JobStatus status = 1;
}

message GetJobResultRequest {
  uint64 job_id = 1;
}

message GetJobResultResponse {
  JobStatus status = 1;
  string result = 2;
  string error_message = 3;
}
